plugin/src/main/kotlin/com/github/gmazzo/gradle/plugins/internal/bindings/JavaHandler.kt

```diff
@@ -4,16 +4,16 @@ import com.github.gmazzo.gradle.plugins.BuildConfigExtension
 import com.github.gmazzo.gradle.plugins.BuildConfigSourceSet
 import com.github.gmazzo.gradle.plugins.generators.BuildConfigJavaGenerator
 import org.gradle.api.Project
+import org.gradle.api.plugins.JavaPluginExtension
 import org.gradle.api.tasks.SourceSet
-import org.gradle.api.tasks.SourceSetContainer
-import org.gradle.kotlin.dsl.the
+import org.gradle.kotlin.dsl.getByName
 
 internal class JavaHandler(
     project: Project,
     private val extension: BuildConfigExtension
 ) : PluginBindingHandler<SourceSet> {
 
-    override val sourceSets = project.the<SourceSetContainer>()
+    override val sourceSets = project.extensions.getByName<JavaPluginExtension>("java").sourceSets
```

Comment:
`JavaPluginExtension` is since Gradle 4.10, but `getSourceSets` is since Gradle 7.1.

Depending on this project's minimum (which I couldn't find, should be in README.md), we might need to condition this call based on `GradleVersion.current()`.

Reply:
Huh, there's also a `project.extensions.getByName<SourceSetContainer>("sourceSets")`, registered by `JvmEcosystemPlugin` (since Gradle 6.7). Not sure which would you prefer. Based on `JavaHandler` I would go with `JavaPluginExtension` if minimum Gradle allows.

Reply:
Hmm... I found the minimum:

https://github.com/gmazzo/gradle-buildconfig-plugin/blob/06c1c0de704277f80a507378b13ec3afab107d68/plugin/src/test/kotlin/com/github/gmazzo/gradle/plugins/BuildConfigPluginTest.kt#L21

So the questions is: Which one do you prefer?
 * bump minimum to 7.1
 * use GradleVersion condition to use conventions in Gradle 7.0
 * use `getByName<SourceSetContainer>("sourceSets")`

Reply:
But `org.gradle.kotlin.dsl.the` is a API function from Gradle's Kotlin DSL. If internally is calling to `convetion` I'd swear they will change the implementation.

Have you tried if this effectively breaks on Gradle 9?

Reply:
Gradle is at 8.2, they're "planning to" remove conventions, not yet, but the deprecation nagging is real and effective now.

That's an interesting point though, since `getByName<SourceSetContainer>("sourceSets")` is a thing, `the<SourceSetContainer>("sourceSets")` must be a thing too. So all we need is a release of this plugin, because, yes, they changed the `the<>()` implementation. The problem is that it's an `inline fun`, so you need to recompile the plugin for it to take effect for your users and they need to upgrade to the new plugin version to be Gradle 8.2 compatible. As far as I see on `main` there's only the changes we need so a `v4.1.2` should resolve this. Sorry for the PR, I only found the root-registered `SourceSetContainer` later.

Reply:
First, thanks for the PRs and contributing to this.

I misted the `inline` factor, and I'm wondering if I should prevent using `the` on any plugin because the backward compatibility changes like this. 

But anyway, I'm just going to release a new patch version now to overcome this limitation for now. 

And thanks again for reporting this 💪

Reply:
No worries, thanks for making action!

I also missed `inline`, a Gradle member called it out: https://github.com/gradle/gradle/issues/25206#issuecomment-1572186680

I personally never liked `the`, it is doing too much, I usually know I'm looking for an extension/extra/etc. so I just explicitly write it out, similar "problem" with `Project.property()`, so I started using `ProviderFactory` wherever possible.